create cluster refactor

mantis-toboggan-md opened the pull request.

eva-vashkevich commented on pkg/capi/components/CCVariables/index.vue:
```diff
@@ -187,65 +322,336 @@ export default {
+            <i
+              class="icon text-primary"
+              :class="{'icon-chevron-down': expanded, 'icon-chevron-up':!expanded}"
+            />Override Defaults
```
should be translation

eva-vashkevich commented on pkg/capi/dialog/AddExampleRepoDialog.vue:
```diff
@@ -0,0 +1,214 @@
+        try {
+          await this.newNamespace.save();
+        } catch (err) {
+          this.$store.dispatch('growl/fromError', { title: 'Error creating namespace', err });
```
Should be translation

eva-vashkevich commented on pkg/capi/dialog/AddExampleRepoDialog.vue:
```diff
@@ -0,0 +1,214 @@
+        cb(true);
+        this.loadBundles();
+      } catch (err) {
+        this.$store.dispatch('growl/fromError', { title: 'Error saving git repo', err });
```
same here

eva-vashkevich commented on pkg/capi/dialog/AddExampleRepoDialog.vue:
```diff
@@ -0,0 +1,214 @@
+
+<template>
+  <div class="dialog-container">
+    <h2>Add Example Clusterclasses</h2>
```
translation

eva-vashkevich commented on pkg/capi/dialog/AddExampleRepoDialog.vue:
```diff
@@ -0,0 +1,214 @@
+      v-if="!gitRepo?.id"
+      class="text-label mb-20"
+    >
+      This will create a fleet git repo resource, <code>turtles-clusterclass-examples</code>, and new namespace in the local cluster,  <code>turtles-clusterclasses</code>, which the example cluster classes will be deployed in. Note that in order for the cluster class resources to be created, you must have already installed the relevant CAPI provider(s).
```
Translation

eva-vashkevich commented on pkg/capi/dialog/AddExampleRepoDialog.vue:
```diff
@@ -0,0 +1,214 @@
+      v-else
+      class="text-label mb-20"
+    >
+      This dialog may be safely closed. Cluster classes will automatically appear on the page when they become available.
```
Translation

eva-vashkevich commented on pkg/capi/dialog/AddExampleRepoDialog.vue:
```diff
@@ -0,0 +1,214 @@
+      >
+        <Checkbox
+          :value="gitRepo?.spec?.paths.includes(path)"
+          :label="key"
```
Translation?

eva-vashkevich commented on pkg/capi/dialog/AddExampleRepoDialog.vue:
```diff
@@ -0,0 +1,214 @@
+          href="#"
+          @click="goToDetail"
+        >
+          View GitRepo deployment details
```
You know what I am about to say

eva-vashkevich commented on pkg/capi/dialog/AddExampleRepoDialog.vue:
```diff
@@ -0,0 +1,214 @@
+          type="button"
+          @click="$emit('close')"
+        >
+          Continue
```
我不明白

eva-vashkevich commented on pkg/capi/edit/cluster.x-k8s.io.cluster/ClusterConfig.vue:
```diff
@@ -275,9 +309,66 @@ export default {
+    },
+
+    isk3s() {
+      return this.clusterClassControlPlane === 'KThreesControlPlaneTemplate';
```
I would prefer if these were constants

eva-vashkevich commented on pkg/capi/edit/cluster.x-k8s.io.cluster/ClusterConfig.vue:
```diff
@@ -406,151 +515,232 @@ export default {
-          :mode="mode"
+        <ClusterClassVariables
+          :will-open="controlPlaneHighlightOpen"
+          :value="value.spec.topology.variables"
```
Maybe we should make it a computed property? in case, at some point, something in that chain becomes undefined

eva-vashkevich commented on pkg/capi/edit/cluster.x-k8s.io.cluster/ClusterConfig.vue:
```diff
@@ -406,151 +515,232 @@ export default {
+      <Accordion
+        class="mt-20"
+        open-initially
+        :title="t(`capi.cluster.section.${formSections.NETWORKING}`)"
```
I am surprised

eva-vashkevich commented on pkg/capi/edit/cluster.x-k8s.io.cluster/WorkerItem.vue:
```diff
@@ -97,23 +107,44 @@ export default {
+    },
+
+    isDeployments() {
+      return this.title.includes('Deployments');
```
I think this should be a translation?

mantis-toboggan-md replied:
Refactored this  a bit - it was a weird way of determining whether the component is being used for pools or deployments.

eva-vashkevich commented on pkg/capi/util/clusterclass-variables.ts:
```diff
@@ -0,0 +1,223 @@
+      }
+      break;
+    case 'string':
+      // if (variable?.metadata?.annotations?.[ANNOTATIONS.SEARCH_TYPE]) {
```
I think we should clean this up

eva-vashkevich commented on pkg/capi/util/clusterclass-variables.ts:
```diff
@@ -0,0 +1,223 @@
+  try {
+    out = createYaml(schemas, 'ccvariable', data, false);
+  } catch (err) {
+    console.error(err);
```
Let's either swallow this error or show it to the user

eva-vashkevich reviewed (changes requested):
There are a couple styling issues I think we can improve:
0. There are some lint issues, especially with spacing
1. I think Workers should be above additional configuration and should have required mark. It is confusing that user can remove the last item (maybe we shouldn't let them), fill in all other required fields and have Create button disabled
2. I think remove button for worker items should be on the right? Even if not, spacing looks awkward, especially on smaller screens.

3. If you try to enter an non-integer value into replica count, 'NaN' is shown instead
4. Override defaults left margin looks a bit awkward? (This is just a preference)
<img width="1312" alt="Screenshot 2025-06-06 at 11 41 09 AM" src="https://github.com/user-attachments/assets/0b41448a-69bf-46fa-a92e-5145cc86d6dc" />
5. There are also some strings in the code that ideally should be constants
6. I think this needs some padding on the right
<img width="1258" alt="Screenshot 2025-06-06 at 3 51 48 PM" src="https://github.com/user-attachments/assets/9230f06a-3de7-4f41-ad48-f4540f2adc18" />
7. This looks wack
<img width="696" alt="Screenshot 2025-06-06 at 3 54 21 PM" src="https://github.com/user-attachments/assets/153bd5ce-2656-433c-ae59-829b0ba28993" />

mantis-toboggan-md commented on pkg/capi/components/CCVariables/index.vue:
```diff
@@ -187,65 +322,336 @@ export default {
+                  :is-machine-scoped="isMachineScoped"
+                  :global-variables="globalVariables"
+                  :validate-required="!machineDeploymentClass && !machinePoolClass"
+                  :cluster-namespace="clusterNamespace"
```
You're right, I changed the name of the variable - updated now

mantis-toboggan-md commented on pkg/capi/components/CCVariables/ResourceVariable.vue:
```diff
@@ -0,0 +1,165 @@
+      try {
+        this.resource = await this.$store.dispatch('management/findAll', { type: this.resourceType, opt});
+      } catch (err) {
+        console.error(err);
```
Since this component is searching for any arbitrary resource type in any namespace in the local cluster I thought errors could be common, and if they can't see resources in a namespace showing the empty dropdown communicates that well enough

mantis-toboggan-md commented on pkg/capi/dialog/AddExampleRepoDialog.vue:
```diff
@@ -0,0 +1,214 @@
+      } catch (err) {
+        this.$store.dispatch('growl/fromError', { title: 'Error saving git repo', err });
+
+        // eslint-disable-next-line node/no-callback-literal
```
I got rid of the lint warning another way (renamed the function)

mantis-toboggan-md commented on pkg/capi/config/capi.ts:
```diff
@@ -38,10 +37,9 @@ export function init($plugin: any, store: any) {
 
   weightType(RANCHER_CAPI.CAPI_CLUSTER, 10, true);
 
-  // Ensure CAPI group appears before the Advanced group
```
I did not full yback out some other changes I was trying out.

eva-vashkevich commented on pkg/capi/edit/cluster.x-k8s.io.cluster/ClusterConfig.vue:
```diff
@@ -375,6 +480,10 @@ export default {
+    },
+
+    openRepoModal() {
+      this.$store.dispatch('management/promptModal', { component: 'AddExampleRepoDialog', modalWidth: '800px' });
```
Is it opening for you correctly? It was and now I am getting 'entry-helpers.js:40 Error: Cannot find module './AddExampleRepoDialog'

mantis-toboggan-md replied:
It works for me. Could be a caching issue if you were on a different branch previously?

eva-vashkevich commented on pkg/capi/edit/cluster.x-k8s.io.cluster/ControlPlaneEndpointSection.vue:
```diff
@@ -92,7 +92,7 @@ export default {
 
 @media screen and (min-width: 1100px) {
```
Should we just remove this block?

eva-vashkevich commented on pkg/capi/edit/cluster.x-k8s.io.cluster/ControlPlaneEndpointSection.vue:
```diff
@@ -102,12 +102,10 @@ export default {
     }
     .col-port {
-        width: 100%;
-        min-width: 120px;
```
any reason we are removing min width?

eva-vashkevich commented on pkg/capi/edit/cluster.x-k8s.io.cluster/WorkerItem.vue:
```diff
@@ -198,20 +237,34 @@ export default {
               type="button"
               :disabled="isView"
               class="btn role-link"
-              :data-testid="`${componentTestid}-remove-item-${idx}`"
```
any reason this is changed?

eva-vashkevich commented on pkg/capi/util/clusterclass-variables.ts:
```diff
@@ -0,0 +1,216 @@
+        if (SIMPLE_TYPES.includes(def.items.type)) {
+          mockSchema.resourceFields[key] = { type: 'array', subtype: def.items.type };
+        } else {
+        // heh
```
I recognize this heh

eva-vashkevich commented on pkg/capi/components/CCVariables/Variable.vue:
```diff
@@ -166,40 +345,91 @@ export default {
-              v-clean-tooltip="validationErrors.join(' ')"
-              class="icon icon-warning"
+      <template #highlight="{toggleOpen}">
+        <label
```
I feel like this was addressed in another PR?

eva-vashkevich reviewed (changes requested):
Ideally, I still think we should add padding to the blue explanation messages to match yellow, but if it is finicky to do, let's not spend more time on css.

I've left a couple comments but feel free to say no

eva-vashkevich commented:
> If that isn't the issue you were referring to a screenshot of what you're seeing would be helpful - they should definitely match! I mostly just don't want to spend ages getting variable css to exactly match our row/col/span-x styling until we're more sure how we want variables styled.

@mantis-toboggan-md Sorry, I was using the wrong term. I should have said right margin. See how the left one starts right where the section border is, but the yellow one doesn't. But up to you if it's something we want to fix
<img width="368" height="166" alt="Screenshot 2025-07-15 at 8 46 11 AM" src="https://github.com/user-attachments/assets/18680fb0-2f53-48c3-a65b-c880630d07fa" />

<img width="557" height="165" alt="Screenshot 2025-07-15 at 8 46 03 AM" src="https://github.com/user-attachments/assets/88d568c4-589b-445f-8ef7-5bc4da8f6c78" />

mantis-toboggan-md commented:
I've updated the gitrepo in the PR description to include better (...working...) example clusterclasses, and the 2 required helm apps with selectors matching these classes.

I see what you mean about the description margins - I've tweaked the style a bit:
<img width="1313" height="1054" alt="Screenshot 2025-08-05 at 8 12 17 AM" src="https://github.com/user-attachments/assets/aab323df-e05c-4d05-a9c1-916504e6136e" />

I also 
- fixed an issue with required variable validation
- added the ability to include a documentation link in variable descriptions
- made some other small style fixes
- fixed component imports in variable util

eva-vashkevich commented on pkg/capi/components/CCVariables/index.vue:
```diff
@@ -1,72 +1,141 @@
+
+    // is  the component being used for top  level cluster variables or machine overrides?
+    isMachineScoped() {
+      return this.machineClassName && this.machineClassType;
```
this needs bang bang

eva-vashkevich replied:
@mantis-toboggan-md still needs !!

mantis-toboggan-md replied:
I am sorry I missed this

eva-vashkevich commented on pkg/capi/edit/cluster.x-k8s.io.cluster/ClusterConfig.vue:
```diff
@@ -406,152 +515,235 @@ export default {
+
       <ClusterClassVariables
-        v-model:value="value.spec.topology.variables"
+        :value="value?.spec?.topology?.variables"
```
let's make this a computed property with a default instead since it is used in multiple places?

eva-vashkevich commented on pkg/capi/edit/cluster.x-k8s.io.cluster/ClusterConfig.vue:
```diff
@@ -406,152 +515,235 @@ export default {
+            >
+              <WorkerItem
+                v-model:value="machinePools"
+                :global-variables="value.spec.topology.variables"
```
let's make it a computed property that checks that everything exists?

eva-vashkevich reviewed (changes requested):
I've left a few more comments.
Additionally:
- I think when switching cluster classes we should clear the fields?
- I haven't been able to reproduce it consistently, but at some point I broke validation. I've left a comment at the part of the code that I think might be related, but it is hard to reproduce
https://github.com/user-attachments/assets/c517e153-3d7b-4c05-823e-cb2e21c009b4
- This warning should be addressed 
<img width="645" height="93" alt="Screenshot 2025-08-06 at 2 59 56 PM" src="https://github.com/user-attachments/assets/5f1879e4-6d51-4057-8c39-403b13258d57" />

- 
Also, styling comments:
- I think we need to add some spacing before the warning icon
<img width="914" height="308" alt="Screenshot 2025-08-06 at 1 44 59 PM" src="https://github.com/user-attachments/assets/24d6da3b-fbde-45cb-8427-903b958b1fd2" />
- I think we should add some space on top before 'Override variables'
<img width="1156" height="409" alt="Screenshot 2025-08-07 at 9 26 56 AM" src="https://github.com/user-attachments/assets/dda3d0f6-fc73-41d7-be8b-eb37623aa071" />

mantis-toboggan-md commented on pkg/capi/components/CCVariables/ResourceVariable.vue:
```diff
@@ -0,0 +1,167 @@
+    // will be passed directly to a labeledselect w/ resource ID as value
+    rules: {
+      type:    Array,
+      default: () => {}
```
Yes you are right. It also wasn't actually wired into the input component! I've fixed both issues now. To test this you can alter one of the examples I provided and removing the default value for `azureClusterIdentity`.

mantis-toboggan-md commented on pkg/capi/components/CCVariables/ResourceVariable.vue:
```diff
@@ -0,0 +1,167 @@
+  components: { LabeledSelect },
+
+  data() {
+    this.fetchResource(this.clusterNamespace);
```
Yes, good call

mantis-toboggan-md commented on pkg/capi/components/CCVariables/VariableHighlight.vue:
```diff
@@ -0,0 +1,297 @@
+      const annotationColor = this.variableDef?.metadata?.annotations?.[ANNOTATIONS.HIGHLIGHT];
+      const searchType = this.variableDef?.metadata?.annotations?.[ANNOTATIONS.SEARCH_TYPE] || '';
+
+      return annotationColor || searchType ? annotationColor || 'info' : '';
```
hmm, yeah, this line is a crime. Refactored

mantis-toboggan-md commented on pkg/capi/components/CCVariables/index.vue:
```diff
@@ -1,72 +1,141 @@
   watch: {
     errorCount: {
-      handler: debounce(function(neu) {
+      handler: throttle(function(neu) {
```
Just trying to reduce the number of events emitted when the component first renders. This watcher gets triggered by each variable with validation

mantis-toboggan-md commented on pkg/capi/components/CCVariables/index.vue:
```diff
@@ -178,22 +313,69 @@ export default {
+
+      // check if next variable is a different component
+      // if it is, force a new line
+      nextVariableDef = variableDefs[i + 1];
```
No; this function is unnecessary since I decided each variable should get its own line. I've removed it now

mantis-toboggan-md commented on pkg/capi/dialog/AddExampleRepoDialog.vue:
```diff
@@ -0,0 +1,219 @@
+        :key="path"
+      >
+        <Checkbox
+          :value="gitRepo?.spec?.paths.includes(path)"
```
No; we are initializing a fresh gitRepo in fetch and this component will never be used to edit an existing gitrepo, so we can trust it

mantis-toboggan-md commented on pkg/capi/dialog/AddExampleRepoDialog.vue:
```diff
@@ -0,0 +1,219 @@
+        v-if="!gitRepo.id"
+        class="btn role-primary"
+        type="button"
+        mode="saveGitRepo"
```
Good catch - The translations for this got removed accidentally when I was rebasing.

eva-vashkevich replied:
I am still confused with how this works now

mantis-toboggan-md commented on pkg/capi/edit/cluster.x-k8s.io.cluster/ClusterConfig.vue:
```diff
@@ -550,210 +574,212 @@ export default {
-        <div class="mt-30">
-          <Checkbox
-            v-model:value="autoImport"
+      <div :key="topology.class">
```
This diff looks crazy but I just added a div here. The key on this div should force all its children to re-initialize when `class` changes.

mantis-toboggan-md commented:
>     * I think when switching cluster classes we should clear the fields?
> 
>     * I haven't been able to reproduce it consistently, but at some point I broke validation. I've left a comment at the part of the code that I think might be related, but it is hard to reproduce

These two appear to be related. I (think) I've solved this validation issue by resetting the cluster object when a new class is chosen, and forcing all the input components to re-render


> 
>     * This warning should be addressed

I have converted machinePools and machineDeployments to computed props with getters/setters

> Also, styling comments:
> 
done & done:
<img width="389" height="389" alt="Screenshot 2025-08-08 at 8 36 45 AM" src="https://github.com/user-attachments/assets/ef263d55-0e44-4e04-bcc5-8e601ed3b4aa" />

mantis-toboggan-md commented on pkg/capi/components/CCVariables/index.vue:
```diff
@@ -178,22 +313,69 @@ export default {
       this.$emit('update:value', out);
     },
 
     updateErrors(isValid) {
```
I THINK the validation issues observed are from hitting the back button while there are errors.

mantis-toboggan-md commented on pkg/capi/components/CCVariables/ResourceVariable.vue:
```diff
@@ -0,0 +1,167 @@
+        opt.namespaced = ns;
+      }
+      try {
+        this.resource = await this.$store.dispatch('management/findAll', { type: this.resourceType, opt });
```
https://github.com/rancher/capi-ui-extension/issues/175

eva-vashkevich commented on pkg/capi/components/CCVariables/VariableHighlight.vue:
```diff
@@ -77,10 +77,13 @@ export default {
+        out = 'info';
+      }
+
+      return out;
```
before we would return '' if annotationColor and searchType are undefined, but now we return undefined. is this intentional?

mantis-toboggan-md replied:
I just thought this looked a bit cleaner. Since both are falsey, they have the same effect here

eva-vashkevich reviewed (changes requested):
I am still seeing the validation issue. I've dm'd you the video
Plus some late feedback.
We should probably disable 'Action' button (which is a confusing name btw) if nothing is selected. And this dialog is missing a cancel button, or some other way to close it without making changes
<img width="1015" height="519" alt="Screenshot 2025-08-08 at 1 24 14 PM" src="https://github.com/user-attachments/assets/0292841b-880d-40c9-b3da-e7aebe7a323a" />

mantis-toboggan-md commented on pkg/capi/edit/cluster.x-k8s.io.cluster/ClusterConfig.vue:
```diff
@@ -268,12 +268,29 @@ export default {
-      return this.value.spec.topology.workers.machinePools;
+    machinePools: {
+      get() {
+        return this.value.spec.topology.workers.machinePools || [];
```
Fair point - added

mantis-toboggan-md commented:
I've removed the `throttle` that seemed to be causing your validation issues. I added a cancel button to the add example repo dialog, and disabled the Create GitRepo button when no paths are selected.

Though it was not causing errors you are correct that all my references to the somewhat nested `paths` field were bad practice - I changed it to a computed prop.

I think the issue with the example repo dialog text was resolved when you shared your screen but let me know if you still see issues there, of course.

eva-vashkevich commented on pkg/capi/dialog/AddExampleRepoDialog.vue:
```diff
@@ -114,11 +125,13 @@ export default {
     addPath(path) {
-      this.gitRepo.spec.paths.push(path);
+      // this.gitRepo.spec.paths.push(path);
+      this.paths.push(path);
```
When we add a path, it's added through here, but removing is done through set.
Even though it works, it's a bit confusing. Up to you if you want to change it, but wanted to mention it

eva-vashkevich reviewed (approved):
Feel free to ignore my comment
